File: packages/sdk/src/frameworks/langgraph-store.ts
// SPDX-License-Identifier: Apache-2.0

import type { MemoryRecord, MemoryType, Scope } from '@memfold/core'
import { type MemfoldLike, findActiveByDedup } from '../common'

/**
 * A stored record in the LangGraph `BaseStore` shape (JS/TS): a namespaced, keyed JSON value with
 * timestamps. See research/08-framework-memory.md.
 */
export interface StoreItem<V = Record<string, unknown>> {
  namespace: string[]
  key: string
  value: V
  createdAt: Date
  updatedAt: Date
}

/** A search result. `SearchItem` in LangGraph terms: an `Item` plus a relevance `score`. */
export interface StoreSearchItem<V = Record<string, unknown>> extends StoreItem<V> {
  score: number | null
}

export interface MemfoldLangGraphStoreOptions {
  /** memfold scope every item is written under. Default `'project'`. */
  scope?: Scope
  /** memfold record type used for items. Default `'fact'`. */
  type?: MemoryType
}

export interface LangGraphSearchOptions {
  query?: string
  limit?: number
}

const SEP = '/'

/**
 * A LangGraph-JS `BaseStore`-shaped adapter backed by memfold. It matches the documented interface
 * (`put` / `get` / `search` / `delete` over `namespace: string[]` + `key`) without importing
 * langchain, so it stays version-agnostic. The hierarchical namespace maps to a memfold
 * `scope_path`, and `namespace + key` becomes the record `dedup_key` (the logical primary key).
 * The value is stored as the JSON body.
 */
export class MemfoldLangGraphStore<V extends Record<string, unknown> = Record<string, unknown>> {
  private readonly scope: Scope
  private readonly type: MemoryType

  constructor(
    private readonly backend: MemfoldLike,
    opts: MemfoldLangGraphStoreOptions = {},
  ) {
    this.scope = opts.scope ?? 'project'
    this.type = opts.type ?? 'fact'
  }

  private nsPath(namespace: string[]): string {
    return namespace.join(SEP)
  }

  private dedupOf(namespace: string[], key: string): string {
    return namespace.length ? `${namespace.join(SEP)}${SEP}${key}` : key
  }

  private toItem(rec: MemoryRecord): StoreItem<V> {
    const namespace = rec.scope_path ? rec.scope_path.split(SEP) : []
    const key =
      rec.scope_path && rec.dedup_key
        ? rec.dedup_key.slice(rec.scope_path.length + SEP.length)
        : (rec.dedup_key ?? '')
    return {
      namespace,
      key,
      value: JSON.parse(rec.body) as V,
      createdAt: new Date(rec.created_at),
      updatedAt: new Date(rec.updated_at),
    }
  }

  private hasPrefix(rec: MemoryRecord, prefix: string[]): boolean {
    const ns = rec.scope_path ? rec.scope_path.split(SEP) : []
    return prefix.every((seg, i) => ns[i] === seg)
  }

  /** Upsert a value at `namespace` / `key`. A prior value at the same address is replaced. */
  async put(namespace: string[], key: string, value: V): Promise<void> {
    const dedup = this.dedupOf(namespace, key)
    const existing = await findActiveByDedup(this.backend, dedup, this.scope)
    if (existing) await this.backend.forget(existing.id)
    await this.backend.add({
      type: this.type,
      scope: this.scope,
      scope_path: this.nsPath(namespace) || null,
      dedup_key: dedup,
      body: JSON.stringify(value),
      provenance: { source: 'agent' },
    })
  }

  async get(namespace: string[], key: string): Promise<StoreItem<V> | null> {
    const rec = await findActiveByDedup(this.backend, this.dedupOf(namespace, key), this.scope)
    return rec ? this.toItem(rec) : null
  }

  async delete(namespace: string[], key: string): Promise<void> {
    const rec = await findActiveByDedup(this.backend, this.dedupOf(namespace, key), this.scope)
    if (rec) await this.backend.forget(rec.id)
  }

  /**
   * Items under `namespacePrefix`. With a `query`, results are ranked by memfold hybrid search;
   * without one, the newest items are returned and `score` is null.
   */
  async search(
    namespacePrefix: string[],
    opts: LangGraphSearchOptions = {},
  ): Promise<StoreSearchItem<V>[]> {
    const limit = opts.limit ?? 10
    if (opts.query) {
      const hits = await this.backend.search(opts.query, {
        scope: this.scope,
        types: [this.type],
        limit: limit * 4,
      })
      return hits
        .filter((h) => this.hasPrefix(h.record, namespacePrefix))
        .slice(0, limit)
        .map((h) => ({ ...this.toItem(h.record), score: h.score }))
    }
    const rows = await this.backend.list({ scope: this.scope, type: this.type, status: 'active' })
    return rows
      .filter((r) => this.hasPrefix(r, namespacePrefix))
      .slice(0, limit)
      .map((r) => ({ ...this.toItem(r), score: null }))
  }
}

File: sdks/python/src/memfold/integrations/langgraph_store.py
# SPDX-License-Identifier: Apache-2.0
"""LangGraph ``BaseStore``-shaped adapter backed by a memfold daemon.

Matches the LangGraph ``BaseStore`` surface (``put`` / ``get`` / ``search`` /
``delete`` / ``list_namespaces`` plus async mirrors) without importing langgraph:
the returned ``Item`` / ``SearchItem`` are local dataclasses that carry the same
fields LangGraph exposes (``namespace``, ``key``, ``value``, ``created_at``,
``updated_at``, and ``score`` on search results).

Mapping onto memfold records:
  * a namespace tuple maps to a record ``scope_path`` (segments joined by ``/``);
  * namespace + key map to a stable ``dedup_key`` so the same key upserts;
  * the value dict is stored as JSON in the record ``body``.
"""

from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass
from typing import Any, Mapping, Sequence

from ..client import MemfoldClient

__all__ = ["MemfoldStore", "Item", "SearchItem"]

_NS_SEP = "/"
_KEY_PREFIX = "lg::"
_KEY_SEP = "::"


@dataclass(slots=True)
class Item:
    """A stored key-value item. Field-compatible with ``langgraph.store.base.Item``."""

    namespace: tuple[str, ...]
    key: str
    value: dict[str, Any]
    created_at: str | None = None
    updated_at: str | None = None


@dataclass(slots=True)
class SearchItem(Item):
    """A search result: an :class:`Item` plus a relevance ``score``."""

    score: float | None = None


def _scope_path(namespace: Sequence[str]) -> str:
    return _NS_SEP.join(namespace)


def _dedup_key(namespace: Sequence[str], key: str) -> str:
    return f"{_KEY_PREFIX}{_scope_path(namespace)}{_KEY_SEP}{key}"


def _key_from_dedup(dedup_key: str | None) -> str:
    if not dedup_key or not dedup_key.startswith(_KEY_PREFIX):
        return dedup_key or ""
    return dedup_key[len(_KEY_PREFIX) :].rsplit(_KEY_SEP, 1)[-1]


class MemfoldStore:
    """A LangGraph ``BaseStore``-shaped store over a memfold daemon.

    Args:
        client: A configured :class:`~memfold.client.MemfoldClient`.
        scope: memfold scope every record is written under (default ``project``).
        type: memfold record type for stored values (default ``fact``).
        tool: provenance tool tag stamped on writes (default ``langgraph``).

    Inject it wherever LangGraph accepts a store. Only the value-store surface is
    implemented; there is no TTL or embedding-index configuration here because the
    daemon owns retrieval.
    """

    supports_ttl: bool = False

    def __init__(
        self,
        client: MemfoldClient,
        *,
        scope: str = "project",
        type: str = "fact",
        tool: str = "langgraph",
    ) -> None:
        self._client = client
        self._scope = scope
        self._type = type
        self._tool = tool

    # -- helpers -----------------------------------------------------------
    def _find(self, namespace: Sequence[str], key: str):
        dedup = _dedup_key(namespace, key)
        rows = self._client.list(
            scope=self._scope, scope_path=_scope_path(namespace), status="active"
        )
        for rec in rows:
            if rec.dedup_key == dedup:
                return rec
        return None

    # -- sync surface ------------------------------------------------------
    def put(
        self,
        namespace: tuple[str, ...],
        key: str,
        value: Mapping[str, Any],
        index: Any = None,
        *,
        ttl: float | None = None,
    ) -> None:
        """Upsert ``value`` at ``namespace`` / ``key``. ``index`` and ``ttl`` are
        accepted for signature compatibility and ignored (the daemon indexes)."""
        body = json.dumps(dict(value))
        existing = self._find(namespace, key)
        if existing is not None:
            self._client.update(existing.id, body=body)
            return
        self._client.add(
            type=self._type,
            scope=self._scope,
            scope_path=_scope_path(namespace),
            body=body,
            dedup_key=_dedup_key(namespace, key),
            provenance={"source": "tool", "tool": self._tool},
        )

    def get(self, namespace: tuple[str, ...], key: str, *, refresh_ttl: bool | None = None) -> Item | None:
        rec = self._find(namespace, key)
        if rec is None:
            return None
        return Item(
            namespace=tuple(namespace),
            key=key,
            value=_load(rec.body),
            created_at=rec.created_at or None,
            updated_at=rec.updated_at or None,
        )

    def delete(self, namespace: tuple[str, ...], key: str) -> None:
        rec = self._find(namespace, key)
        if rec is not None:
            self._client.forget(rec.id)

    def search(
        self,
        namespace_prefix: tuple[str, ...],
        /,
        *,
        query: str | None = None,
        filter: Mapping[str, Any] | None = None,
        limit: int = 10,
        offset: int = 0,
        refresh_ttl: bool | None = None,
    ) -> list[SearchItem]:
        path = _scope_path(namespace_prefix)
        if query:
            hits = self._client.search(
                query, scope=self._scope, scope_path=path, limit=limit + offset
            )
            items = [
                SearchItem(
                    namespace=tuple(namespace_prefix),
                    key=_key_from_dedup(h.record.dedup_key),
                    value=_load(h.record.body),
                    created_at=h.record.created_at or None,
                    updated_at=h.record.updated_at or None,
                    score=h.score,
                )
                for h in hits
            ]
        else:
            rows = self._client.list(scope=self._scope, scope_path=path, status="active")
            items = [
                SearchItem(
                    namespace=tuple(namespace_prefix),
                    key=_key_from_dedup(rec.dedup_key),
                    value=_load(rec.body),
                    created_at=rec.created_at or None,
                    updated_at=rec.updated_at or None,
                    score=None,
                )
                for rec in rows
            ]
        if filter:
            items = [it for it in items if _matches(it.value, filter)]
        return items[offset : offset + limit]

    def list_namespaces(
        self,
        *,
        prefix: tuple[str, ...] | None = None,
        suffix: tuple[str, ...] | None = None,
        max_depth: int | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> list[tuple[str, ...]]:
        rows = self._client.list(scope=self._scope, status="active")
        seen: list[tuple[str, ...]] = []
        for rec in rows:
            if not rec.scope_path:
                continue
            ns = tuple(rec.scope_path.split(_NS_SEP))
            if max_depth is not None:
                ns = ns[:max_depth]
            if prefix and ns[: len(prefix)] != tuple(prefix):
                continue
            if suffix and ns[-len(suffix) :] != tuple(suffix):
                continue
            if ns not in seen:
                seen.append(ns)
        return seen[offset : offset + limit]

    # -- async mirrors -----------------------------------------------------
    async def aput(self, *args: Any, **kwargs: Any) -> None:
        return await asyncio.to_thread(self.put, *args, **kwargs)

    async def aget(self, *args: Any, **kwargs: Any) -> Item | None:
        return await asyncio.to_thread(self.get, *args, **kwargs)

    async def adelete(self, *args: Any, **kwargs: Any) -> None:
        return await asyncio.to_thread(self.delete, *args, **kwargs)

    async def asearch(self, *args: Any, **kwargs: Any) -> list[SearchItem]:
        return await asyncio.to_thread(self.search, *args, **kwargs)

    async def alist_namespaces(self, *args: Any, **kwargs: Any) -> list[tuple[str, ...]]:
        return await asyncio.to_thread(self.list_namespaces, *args, **kwargs)


def _load(body: str) -> dict[str, Any]:
    try:
        loaded = json.loads(body)
    except (ValueError, TypeError):
        return {"value": body}
    return loaded if isinstance(loaded, dict) else {"value": loaded}


def _matches(value: Mapping[str, Any], filter: Mapping[str, Any]) -> bool:
    return all(value.get(k) == v for k, v in filter.items())
